src/virt.c

```diff
@@ -16,998 +17,900 @@
-            virDomainPtr dom, const char *type)
-{
-    submit (dom, type, NULL, &(value_t) { .derive = (derive_t) value }, 1);
+static int virt2_submit(const virt2_config_t *cfg, const VMInfo *info,
+                        const char *type, const char *type_instance,
+                        value_t *values, size_t values_len) {
+  value_list_t vl = VALUE_LIST_INIT;
+
+  sstrncpy(vl.plugin, PLUGIN_NAME, sizeof(vl.plugin));
+  virt2_value_list_set_plugin_instance(&vl, info, cfg);
+  virt2_value_list_set_host(&vl, info, cfg);
+  sstrncpy(vl.type, type, sizeof(vl.type));
+  sstrncpy(vl.type_instance, type_instance, sizeof(vl.type_instance));
+
+  vl.values = values;
+  vl.values_len = values_len;
+
+  plugin_dispatch_values(&vl);
+  return 0;
 }
 
-static void
-vcpu_submit (derive_t value,
-             virDomainPtr dom, int vcpu_nr, const char *type)
-{
+static int virt2_dispatch_cpu(virt2_instance_t *inst, const VMInfo *vm) {
+  value_t val;
+
+  val.derive = vm->info.cpuTime;
+  virt2_submit(inst->conf, vm, "virt_cpu_total", "", &val, 1);
+  // TODO: cpu.user, cpu.sys, cpu.total
+
+  for (size_t j = 0; j < vm->vcpu.nstats; j++) {
     char type_instance[DATA_MAX_NAME_LEN];
+    ssnprintf(type_instance, sizeof(type_instance), "%zu", j);
+    const VCpuStats *stats =
+        (vm->vcpu.xstats) ? vm->vcpu.xstats : vm->vcpu.stats;
+    val.derive = stats[j].time;
+    virt2_submit(inst->conf, vm, "virt_vcpu", type_instance, &val, 1);
+  }
+
+  return 0;
+}
 
-    ssnprintf (type_instance, sizeof (type_instance), "%d", vcpu_nr);
+static int virt2_dispatch_memory(virt2_instance_t *inst, const VMInfo *vm) {
+  value_t val;
+  val.gauge = vm->info.memory * 1024;
+  virt2_submit(inst->conf, vm, "memory", "total", &val, 1);
+  for (int j = 0; j < vm->memstats_count; j++) {
+    static const char *tags[] = {"swap_in",        "swap_out", "major_fault",
+                                 "minor_fault",    "unused",   "available",
+                                 "actual_balloon", "rss"};
+    if ((vm->memstats[j].tag < 0) ||
+        (vm->memstats[j].tag >= STATIC_ARRAY_SIZE(tags))) {
+      ERROR(PLUGIN_NAME " plugin#%s: unknown tag %i", inst->tag,
+            vm->memstats[j].tag);
+      continue;
+    }
+    val.gauge = vm->memstats[j].val * 1024;
+    virt2_submit(inst->conf, vm, "memory", tags[vm->memstats[j].tag], &val, 1);
+  }
 
-    submit (dom, type, type_instance, &(value_t) { .derive = value }, 1);
+  return 0;
 }
 
-static void
-submit_derive2 (const char *type, derive_t v0, derive_t v1,
-             virDomainPtr dom, const char *devname)
-{
-    value_t values[] = {
-        { .derive = v0 },
-        { .derive = v1 },
-    };
-
-    submit (dom, type, devname, values, STATIC_ARRAY_SIZE (values));
-} /* void submit_derive2 */
-
-static int
-lv_init (void)
-{
-    if (virInitialize () != 0)
-        return -1;
-    else
-        return 0;
+static int virt2_dispatch_block(virt2_instance_t *inst, const VMInfo *vm) {
+  value_t vals[2];
```

Comment:
Magic number. You could use an enum to specify array size and particular indexes i.e BLOCK_STATS_RD_IDX = 0, BLOCK_STATS_WR_IDX = 1,  BLOCK_STATS_NR = 2. Such approach could be used with other metrics like interface stats.